cli/democluster: TestTransientClusterMultitenant failed

cli/democluster.TestTransientClusterMultitenant [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10967894?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10967894?buildTab=artifacts#/) on release-23.1 @ [8897f4d13a030381997ff30aeaf633a9f8e87340](https://github.com/cockroachdb/cockroach/commits/8897f4d13a030381997ff30aeaf633a9f8e87340):


```
=== RUN   TestTransientClusterMultitenant
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/1937dc802442f15720877ed362f2998f/logTestTransientClusterMultitenant1865493125
    test_log_scope.go:79: use -show-logs to present logs inline
    demo_cluster_test.go:298: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/cli/democluster/demo_cluster_test.go:298
        	Error:      	Received unexpected error:
        	            	server stop before successful start
        	            	(1) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/server.(*channelOrchestrator).startControlledServer.func5.1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/server/server_controller_channel_orchestrator.go:307
        	            	  | github.com/cockroachdb/cockroach/pkg/server.(*channelOrchestrator).startControlledServer.func5
        	            	  | 	github.com/cockroachdb/cockroach/pkg/server/server_controller_channel_orchestrator.go:401
        	            	  | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2
        	            	  | 	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:470
        	            	  | runtime.goexit
        	            	  | 	GOROOT/src/runtime/asm_amd64.s:1594
        	            	Wraps: (2) server stop before successful start
        	            	Error types: (1) *withstack.withStack (2) *errutil.leafError
        	Test:       	TestTransientClusterMultitenant
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/1937dc802442f15720877ed362f2998f/logTestTransientClusterMultitenant1865493125
--- FAIL: TestTransientClusterMultitenant (19.67s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations @cockroachdb/server @cockroachdb/server
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTransientClusterMultitenant.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-29918

--- a/pkg/server/server_controller_orchestration.go
+++ b/pkg/server/server_controller_orchestration.go
@@ -96,6 +96,9 @@ func (c *serverController) startInitialSecondaryTenantServers(
 func (c *serverController) scanTenantsForRunnableServices(
 	ctx context.Context, ie isql.Executor,
 ) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
 	// The list of tenants that should have a running server.
 	reqTenants, err := c.getExpectedRunningTenants(ctx, ie)
 	if err != nil {
@@ -107,9 +110,6 @@ func (c *serverController) scanTenantsForRunnableServices(
 	for _, name := range reqTenants {
 		nameLookup[name] = struct{}{}
 	}
-
-	c.mu.Lock()
-	defer c.mu.Unlock()
 
 	// First check if there are any servers that shouldn't be running
 	// right now.

